Deprecation of Bigarray.*.map_file and introduction of Unix.map_file, continued

xavierleroy opened the pull request:
This is a follow-up to GPR #997 with a different implementation that creates fewer dependencies.

We add to the runtime a `byterun/bigarray.c` file that contains the bigarray creation functions that used to be in `otherlibs/bigarray/bigarray_stubs.c`.  In the latter file we keep all primitives needed to implement the Bigarray interface.  The functions in the new `byterun/bigarray.c` make it possible to create bigarrays both from the bigarray library and from the unix library.

The header file `bigarray.h` moves to `byterun/caml/bigarray.h` accordingly.

The `map_file` implementations move to `otherlibs/unix/mmap.c` and `otherlibs/win32unix/mmap.c`.  Some bigarray allocation code shared between the two implementations is put in `otherlibs/unix/mmap_ba.c`.

Through a couple of `#ifdef`, the `map_file` implementations can also be compiled from within `otherlibs/bigarray` with the same semantics they had in 4.04.

As a consequence, the bigarray library still contains a standalone, Unix-independent implementation of `Bigarray.*.map_file`; the only difference with 4.04 is that it is marked deprecated.

Current status: compiled and lightly tested under Unix.  Win32 implementation neither compiled nor tested.  I'm submitting it as a GPR so that it gets some CI.

xavierleroy committed:
Deprecation of Bigarray.*.map_file and introduction of Unix.map_file, continued

This is a follow-up to GPR #997 with a different implementation that creates fewer dependencies.

We add to the runtime a `byterun/bigarray.c` file that contains the bigarray creation functions that used to be in `otherlibs/bigarray/bigarray_stubs.c`.  In the latter file we keep all primitives needed to implement the Bigarray interface.  The functions in the new `byterun/bigarray.c` make it possible to create bigarrays both from the bigarray library and from the unix library.

The header file `bigarray.h` moves to `byterun/caml/bigarray.h` accordingly.

The `map_file` implementations move to `otherlibs/unix/mmap.c` and `otherlibs/win32unix/mmap.c`.  Some bigarray allocation code shared between the two implementations is put in `otherlibs/unix/mmap_ba.c`.

Through a couple of `#ifdef`, the `map_file` implementations can also be compiled from within `otherlibs/bigarray` with the same semantics they had in 4.04.

As a consequence, the bigarray library still contains a standalone, Unix-independent implementation of `Bigarray.*.map_file`; the only difference with 4.04 is that it is marked deprecated.

Current status: compiled and lightly tested under Unix.  Win32 implementation neither compiled nor tested.

xavierleroy committed:
Possibly wrong #include + update dependencies

xavierleroy committed:
Define CAML_INTERNALS to see more of caml/bigarray.h

gasche commented:
From appveyor:

```
Running tests from 'tests/unboxed-primitive-args' ...
test_common.c(17) : fatal error C1083: Cannot open include file: 'bigarray.h': No such file or directory
```

xavierleroy committed:
Update tests/unboxed-primitive-args: bigarray.h is now in caml/

xavierleroy commented:
Thanks for the ping.  Pushed a trivial fix for the failing test.

xavierleroy committed:
win32unix/mmap.c: error reporting was all wrong, attempting a fix

xavierleroy committed:
map_file: fix a C name clash and add more tests

caml_unix_map_file and related functions were defined both in the stub lib for Unix and in the stub lib for Bigarray.  This caused strange behavior when both Unix and Bigarray were linked, as found by the extra tests.

The extra tests (all two of them) check that Unix.map_file raises reasonable Unix_error exceptions in common error cases.

xavierleroy committed:
More tweaks on the Unix.map_file error test

gasche commented on testsuite/tests/lib-bigarray-file/mapfile.ml:
```diff
@@ -86,7 +86,32 @@ let tests () =
+      with
+      | Unix.Unix_error((Unix.EACCES | Unix.EPERM), _, _) -> true
+      | Unix.Unix_error(err, _, _) ->
+          Printf.eprintf "Unexpected error %s\n" (Unix.error_message err);
```
It would be nice to include a flush `%!` to the format to avoid delayed printing. (Same below)

gasche commented:
It looks like the test still fails on Windows:

```
Running tests from 'tests/lib-bigarray-file' ...
 ... testing 'mapfile.ml': ocamlcUnexpected error Invalid argument
*** Bad result (map_file errors, test 1)
************* TEST FAILED ****************
make[7]: *** [/cygdrive/c/projects/ocaml/testsuite/makefiles/Makefile.several:114: run-file] Error 2
 => failed
make[2]: Entering directory '/cygdrive/c/projects/ocaml/testsuite'
```

(@xavierleroy: I'm writing this tautologic message because I'm not sure you have seen the last failure; I myself don't get notifications when the CI results are available, which would be convenient, so I have to resort to polling; given that I just did a polling cycle I thought you could be interested in a delayed notification. Feel free to grumble otherwise.)

ghost commented:
@xavierleroy I think we need to reference at least one stub from `byterun/bigarray.c` in the stdlib, otherwise we won't be able to dynlink bigarray in native code. We had the same problem with the compiler libraries in the past: they use stubs from `byterun/terminfo.c` but since nothing in the stdlib references them, `terminfo.o` is never statically linked in, even with `-linkall`

xavierleroy commented:
@gasche : like you, I poll occasionally.  I'm on the Win32 test failure.

@diml : I ran into the issue you mention.  Currently, it is avoided by forcing the registration of `caml_ba_ops` (the bigarray custom operations) in `caml_init_custom_operations`, with the extra benefit that there is no longer any need for initialization code in the Bigarray library.

xavierleroy committed:
Unix.map_file: fix error reporting in the win32 implementation

Also: cosmetic improvement to corresponding test.

xavierleroy commented:
All right, it seems that CI is happy (the Travis failure is because of no Changes entry, despite the label to that effect), so I turned off the fasten seat belts sign ^H^H^H^H^H^H I removed the work-in-progress label to indicate that this PR is ready for code review.  (Yay.)  I'm sorry that it turned out a little more complicated than I hoped.  On the other hand, some of the `#ifdef` ugliness will go away when `Bigarray.*.map_file` is removed for good.

gasche commented:
(It would definitely make sense to add the number of the present GPR to the GPR#997 change entry, so that people can have the full picture.)

xavierleroy committed:
Update Changes for GPR#1077 (Unix.map_file)

xavierleroy commented:
Fair enough, in the end I updated the Changes file.

gasche commented:
@diml, do you think it would be possible for you (or someone else) to review this proposed change in the following days? This is the only remaining blocker that I am aware of for Camomile, which means that after this is merged (or, if this takes too long, after #997 is reverted as a first step), I could ask @lefessan to update the `opam-builder` bot to a more recent version of the 4.05 branch.

ghost commented:
@gasche sure, I was planning to look at this today or tomorrow

gasche commented:
Thanks!

diml commented on otherlibs/unix/mmap_ba.c:
```diff
@@ -0,0 +1,91 @@
+  Assert((flags & CAML_BA_KIND_MASK) <= CAML_BA_CHAR);
+  for (i = 0; i < num_dims; i++) dimcopy[i] = dim[i];
+  asize = SIZEOF_BA_ARRAY + num_dims * sizeof(intnat);
+  res = caml_alloc_custom(&caml_ba_mapped_ops, asize, 0, 1);
```
What is the motivation for using `0, 1` rather than `size, CAML_BA_MAX_MEMORY` as before? Isn't that bad if `shared=false` for instance?

diml replied:
BTW, I never remember how these values are interpreted, so `0, 1` might be the right setting, but I would think it's safer to be conservative for the upcoming release

xavierleroy replied:
The old code used `0` and `CAML_BA_MAX_MEMORY` in this case (i.e. `size` is 0 for mmaped bigarrays), so the new code behaves like the old code.  Whether it is the right thing to do is unclear.

diml replied:
Indeed, I missed that the size was 0 in the old code

diml commented on otherlibs/win32unix/mmap.c:
```diff
@@ -134,3 +157,28 @@ void caml_ba_unmap_file(void * addr, uintnat len)
+   It is included for backward compatibility with the old
+   Bigarray.*.map_file implementation.  */
+
+static void caml_ba_sys_error(void)
```
This function seems to be unused

xavierleroy replied:
You're correct, wrong cut and paste from ../unix/mmap.c.  Fixed.

shindere commented:
@xavierleroy:

In byterun/bigarray.c:

- The hearder looks wrong: 200 -> 2017, Rocquencourt -> Paris
  (perhaps Manuel's name also?)

- The file uses both Assert and CAMLassert, I am wondering in
  which context each of them should be used

Incidentally, in byterun/caml/misc.h one reads:

#ifndef CAML_AVOID_CONFLICTS
#define Assert CAMLassert
#endif

And it seems it's the only place in the repo that mentions this
CAML_AVOID_CONFLICTS macro.

And finally, before merging, wouldn't it be worth sending this PR to precheck?

xavierleroy committed:
Wrong error reporting in the IN_OCAML_BIGARRAY legacy case

xavierleroy committed:
bigarray.c: Assert -> CAMLassert

xavierleroy commented:
@shindere : OK for `Assert` that should be `CAMLassert`, fixed in the latest commit.

Concerning the header, it is the same as for otherlibs/bigarray/bigarray_stubs.c, from which byterun/bigarray.c was excerpted.  Yes, the bigarray_stubs.c dates back to 2000, was written with the help of Manuel Serrano, and I was located in Rocquencourt at this time.  We don't have a policy of updating file headers, so why not keep the original header?

shindere commented:
Thanks, @xavierleroy. I'm very fine with anything. Just wanted to make
sure it was not a mistake.

Re Assert: shall we get rid of CAML_AVOID_CONFLICTS and of Assert
itself and consistently replace it by CAMLassert everywhere?

damiendoligez commented:
>Re Assert: shall we get rid of CAML_AVOID_CONFLICTS and of Assert itself and consistently replace it by CAMLassert everywhere?

That would be a good thing, but of course not in this PR.

shindere commented:
Damien Doligez (2017/03/09 05:08 -0800):
> >Re Assert: shall we get rid of CAML_AVOID_CONFLICTS and of Assert itself and consistently replace it by CAMLassert everywhere?
> 
> That would be a good thing, but of course not in this PR.

Sure. Will do it in a different PR. Can Assert be removed completely,
all should it rather be moved to compatibility.h?

I would personnally remove it because as far as I know it is not
documented so user-code is not supposed to use it.

damiendoligez commented:
Yes, remove it completely.

ghost commented:
The windows CI is failing with:

```
../win32unix/mmap.c(168) : error C2371: 'caml_ba_sys_error' : redefinition; different basic types
```

Otherwise the I read the changes and the patch looks correct to me.

xavierleroy committed:
Missing declaration of caml_ba_sys_error()

Follow-up to e763c83

gasche commented:
CI systems force us to realize that we work to make the machine happy, and not the other way around.

xavierleroy committed:
Wrong .depend, shoot again

xavierleroy committed:
Replace Assert with CAMLassert

xavierleroy commented:
One more conflict resolved.  One more rebasing done.  I'm getting tired.  Can we please merge? pretty please?

gasche commented:
The CI is "all green" and the merge target is trunk. I think it's an easy choice to merge -- let's do it right now. But the real question is whether you also want to merge this in 4.05?

shindere commented:
This broke CI because the defined MAP_FILE clashes with a macro defined in
sys/mman.h.

Fix pushed to trunk: a01cfa2b5af1e337eee8e796c61d58a677a3d985

xavierleroy commented:
@shindere : thanks for fixing this problem that I should have seen coming!

@gasche : I don't know if I want to backport this to 4.05 (it will take a bit of work, now that we've done the revert on 4.05).  I'd like some more discussion on caml-devel and a consensus.

shindere commented:
Xavier Leroy (2017/03/14 12:26 -0700):
> @shindere : thanks for fixing this problem that I should have seen
> coming!

My pleasure!

shindere commented:
Other fix pushed to trunk: a6d77053692017224565d48fe62bdeff291d0781
(missing inclusion of caml_copy_string in otherlibs/win32unix/mmap.c).
